runtime/parachains/src/paras.rs

```diff
@@ -0,0 +1,1103 @@
+//! only occur at session boundaries.
+
+use sp_std::prelude::*;
+use sp_std::marker::PhantomData;
+use sp_runtime::traits::One;
+use primitives::{
+	parachain::{ValidatorId, Id as ParaId, ValidationCode, HeadData},
+};
+use frame_support::{
+	decl_storage, decl_module, decl_error,
+	traits::Get,
+	weights::Weight,
+};
+use codec::{Encode, Decode};
+use crate::configuration;
+
+#[cfg(feature = "std")]
+use serde::{Serialize, Deserialize};
+
+pub trait Trait: system::Trait + configuration::Trait { }
+
+/// Metadata used to track previous parachain validation code that we keep in
+/// the state.
+#[derive(Default, Encode, Decode)]
+#[cfg_attr(test, derive(Debug, Clone, PartialEq))]
+pub struct ParaPastCodeMeta<N> {
+	// Block numbers where the code was "technically" replaced and the block number at
+	// which the code was actually replaced. These can be used as indices
+	// into the `PastCode` map along with the `ParaId` to fetch the code itself.
+	upgrade_times: Vec<(N, N)>,
+	// This tracks the highest pruned code-replacement, if any.
+	last_pruned: Option<N>,
+}
+
+#[cfg_attr(test, derive(Debug, PartialEq))]
+enum UseCodeAt<N> {
+	// Use the current code.
+	Current,
+	// Use the code that was replaced at the given block number.
+	ReplacedAt(N),
+}
+
+impl<N: Ord + Copy> ParaPastCodeMeta<N> {
+	// note a replacement has occurred at a given block number.
+	fn note_replacement(&mut self, at: N, included_at: N) {
+		self.upgrade_times.insert(0, (at, included_at))
+	}
+
+	// Yields the block number of the code that should be used for validating at
+	// the given block number.
+	//
+	// a return value of `None` means that there is no code we are aware of that
+	// should be used to validate at the given height.
+	#[allow(unused)]
+	fn code_at(&self, at: N) -> Option<UseCodeAt<N>> {
+		// The `PastCode` map stores the code which was replaced at `t`.
+		let end_position = self.upgrade_times.iter().position(|&t| t.0 < at);
+		if let Some(end_position) = end_position {
+			Some(if end_position != 0 {
+				// `end_position` gives us the replacement time where the code used at `at`
+				// was set. But that code has been replaced: `end_position - 1` yields
+				// that index.
+				UseCodeAt::ReplacedAt(self.upgrade_times[end_position - 1].0)
+			} else {
+				// the most recent tracked replacement is before `at`.
+				// this means that the code put in place then (i.e. the current code)
+				// is correct for validating at `at`.
+				UseCodeAt::Current
+			})
+		} else {
+			if self.last_pruned.as_ref().map_or(true, |&n| n < at) {
+				// Our `last_pruned` is before `at`, so we still have the code!
+				// but no code upgrade entries found before the `at` parameter.
+				//
+				// this means one of two things is true:
+				// 1. there are no non-pruned upgrade logs. in this case use `Current`
+				// 2. there are non-pruned upgrade logs all after `at`.
+				//    in this case use the oldest upgrade log.
```

Comment:
(I am writing this after rewriting @coriolinus' suggestion to undo the reversal. I find the code clearer now)

The easiest case to explain is the one where we have done zero code replacements ever. In that case, we know that the `Current` code is the right one to use.

Another case is where we have a situation like this:

```rust
ParaPastCodeMeta {
    upgrade_times: vec![(10, 12), (20, 25), ...],
    last_pruned: Some(5),
}
```

Let's look at `code_at(6)`. The earliest tracked replacement we have is at `10`, so does that mean we should return `ReplacedAt(10)`? No, because there might have been another replacement at e.g. `8` which was pruned. So we need to compare against the `last_pruned` member. In this case `last_pruned` is `5`, so since 6 falls between 5 and 10 we know that 6 was validated with the code replaced at block 10, from the para's perspective.

However, `code_at(5)` was validated using the code that was replaced at `5`, so returning `ReplacedAt(10)` would be wrong. We'd want to return `ReplacedAt(5)`, but we are aware that this was pruned, and thus can only return `None`.